gvfs-helper.c

```diff
@@ -0,0 +1,2074 @@
+ * does not do 401s (or manage creds), we have to reload the main Git
+ * server creds first.
+ *
+ * That is, we NEVER reject them directly because we never filled them.
+ */
+static void refresh_cache_server_creds(void)
+{
+	credential_clear(&gh__global.cache_creds);
+
+	refresh_main_creds();
+	synthesize_cache_server_creds();
+}
+
+/*
+ * We NEVER approve cache-server creds directly because we never directly
+ * filled them.  However, we should be able to infer that the main ones
+ * are valid and can approve them if necessary.
+ */
+static void approve_cache_server_creds(void)
+{
+	approve_main_creds();
+}
+
+/*
+ * Select the ODB directory where we will write objects that we
+ * download.  If no product-specific ODB is defined, use the
+ * local ODB (in ".git/objects").
+ */
+static void select_odb(void)
+{
+	const char *p_odb = the_repository->objects->odb->path;
+
+	strbuf_init(&gh__global.buf_odb_path, 0);
+
+	switch (gh__global.chosen_product_mode) {
+	default:                      /* should not happen */
+	case GH__PRODUCT_MODE__AUTO:  /* should not happen */
+	case GH__PRODUCT_MODE__NONE:
+		break;
+
+	case GH__PRODUCT_MODE__VFS:
+		if (gh__cmd_opts.vfs_odb_path)
+			p_odb = gh__cmd_opts.vfs_odb_path;
+		break;
+
+	case GH__PRODUCT_MODE__SCALAR:
+		if (gh__cmd_opts.scalar_odb_path)
+			p_odb = gh__cmd_opts.scalar_odb_path;
+		break;
+	}
+
+	strbuf_addstr(&gh__global.buf_odb_path, p_odb);
+}
+
+/*
+ * Create a tempfile to stream the packfile into.
+ *
+ * We create a tempfile in the chosen ODB directory and let CURL
+ * automatically stream data to the file.  If successful, we can
+ * later rename it to a proper .pack and run "git index-pack" on
+ * it to create the corresponding .idx file.
+ *
+ * TODO I would rather to just stream the packfile directly into
+ * TODO "git index-pack --stdin" (and save some I/O) because it
+ * TODO will automatically take care of the rename of both files
+ * TODO and any other cleanup.  BUT INDEX-PACK WILL ONLY WRITE
+ * TODO TO THE PRIMARY ODB -- it will not write into the alternates
+ * TODO (this is considered bad form).  So we would need to add
+ * TODO an option to index-pack to handle this.  I don't want to
+ * TODO deal with this issue right now.
+ *
+ * TODO Consider using lockfile for this rather than naked tempfile.
+ */
+static struct tempfile *create_tempfile_for_packfile(void)
+{
+	static unsigned int nth = 0;
+	static struct timeval tv = {0};
+	static struct tm tm = {0};
+	static time_t secs = 0;
+	static char tbuf[32] = {0};
+
+	struct tempfile *tempfile = NULL;
+	struct strbuf buf_path = STRBUF_INIT;
+
+	if (!nth) {
+		gettimeofday(&tv, NULL);
+		secs = tv.tv_sec;
+		gmtime_r(&secs, &tm);
+
+		xsnprintf(tbuf, sizeof(tbuf), "%4d%02d%02d-%02d%02d%02d-%06ld",
+			  tm.tm_year + 1900, tm.tm_mon + 1, tm.tm_mday,
+			  tm.tm_hour, tm.tm_min, tm.tm_sec,
+			  (long)tv.tv_usec);
+	}
+
+	// TODO should this be in the "<ODB>/pack/tempPacks/"
```

Comment:
Yes, the `tempPacks` dir would more closely match the behavior of `scalar prefetch --folders` _and_ we would avoid issues with hardware failure before a full flush during indexing.